File: src/swift/4_BehavioralPatterns/12_Visitor/Exercise.swift
import Foundation
import XCTest

protocol ExpressionVisitor
{
  func accept(_ value: Value)
  func accept(_ ae: AdditionExpression)
  func accept(_ me: MultiplicationExpression)
}

protocol Expression
{
  func visit(_ ev: ExpressionVisitor)
}

class Value : Expression
{
  let value: Int
  init(_ value: Int)
  {
    self.value = value
  }
  func visit(_ ev: ExpressionVisitor)
  {
    ev.accept(self)
  }
}

class AdditionExpression : Expression
{
  let lhs, rhs: Expression
  init(_ lhs: Expression, _ rhs: Expression)
  {
    self.lhs = lhs
    self.rhs = rhs
  }
  func visit(_ ev: ExpressionVisitor)
  {
    ev.accept(self)
  }
}

class MultiplicationExpression : Expression
{
  let lhs, rhs: Expression
  init(_ lhs: Expression, _ rhs: Expression)
  {
    self.lhs = lhs
    self.rhs = rhs
  }
  func visit(_ ev: ExpressionVisitor)
  {
    ev.accept(self)
  }
}

class ExpressionPrinter :
  ExpressionVisitor, CustomStringConvertible
{
  private var buffer = ""

  func accept(_ value: Value)
  {
    buffer.append(String(value.value))
  }

  func accept(_ ae: AdditionExpression)
  {
    buffer.append("(")
    ae.lhs.visit(self)
    buffer.append("+")
    ae.rhs.visit(self)
    buffer.append(")")
  }

  func accept(_ me: MultiplicationExpression)
  {
    me.lhs.visit(self)
    buffer.append("*")
    me.rhs.visit(self)
  }

  var description: String
  {
    return buffer
  }
}

class UMBaseTestCase : XCTestCase {}

class Evaluate: UMBaseTestCase
{
  func simpleAddition()
  {
    let simple = AdditionExpression(
      Value(2), Value(3)
    )
    let ep = ExpressionPrinter()
    ep.accept(simple)
    XCTAssertEqual("(2+3)", ep.description)
  }

  func productOfAdditionAndValue()
  {
    let expr = MultiplicationExpression(
      AdditionExpression(Value(2), Value(3)),
      Value(4)
    )
    let ep = ExpressionPrinter()
    ep.accept(expr)
    XCTAssertEqual("(2+3)*4", ep.description)
  }
}

extension Evaluate
{
  static var allTests : [(String, (Evaluate) -> () throws -> Void)]
  {
    return [
      ("simpleAddition", simpleAddition),
      ("productOfAdditionAndValue", productOfAdditionAndValue)
    ]
  }
}

func main()
{
  XCTMain([testCase(Evaluate.allTests)])
}

main()

File: src/csharp/4_BehavioralPatterns/12_Visitor/Exercise.cs
using System;
using System.Activities.Expressions;
using System.Activities.Validation;
using System.Text;
using NUnit.Framework;

namespace DotNetDesignPatternDemos.Behavioral.Visitor
{
  namespace Coding.Exercise
  {
    public abstract class ExpressionVisitor
    {
      public abstract void Visit(Value value);
      public abstract void Visit(AdditionExpression ae);
      public abstract void Visit(MultiplicationExpression me);
    }

    public abstract class Expression
    {
      public abstract void Accept(ExpressionVisitor ev);
    }

    public class Value : Expression
    {
      public readonly int TheValue;

      public Value(int value)
      {
        TheValue = value;
      }

      public override void Accept(ExpressionVisitor ev)
      {
        ev.Visit(this);
      }
    }

    public class AdditionExpression : Expression
    {
      public readonly Expression LHS, RHS;

      public AdditionExpression(Expression lhs, Expression rhs)
      {
        LHS = lhs;
        RHS = rhs;
      }

      public override void Accept(ExpressionVisitor ev)
      {
        ev.Visit(this);
      }
    }

    public class MultiplicationExpression : Expression
    {
      public readonly Expression LHS, RHS;

      public MultiplicationExpression(Expression lhs, Expression rhs)
      {
        LHS = lhs;
        RHS = rhs;
      }

      public override void Accept(ExpressionVisitor ev)
      {
        ev.Visit(this);
      }
    }

    public class ExpressionPrinter : ExpressionVisitor
    {
      private StringBuilder sb = new StringBuilder();

      public override void Visit(Value value)
      {
        sb.Append(value.TheValue);
      }

      public override void Visit(AdditionExpression ae)
      {
        sb.Append("(");
        ae.LHS.Accept(this);
        sb.Append("+");
        ae.RHS.Accept(this);
        sb.Append(")");
      }

      public override void Visit(MultiplicationExpression me)
      {
        me.LHS.Accept(this);
        sb.Append("*");
        me.RHS.Accept(this);
      }

      public override string ToString()
      {
        return sb.ToString();
      }
    }
  }

  namespace Coding.Exercise.Tests
  {
    [TestFixture]
    public class Tests
    {
      [Test]
      public void SimpleAddition()
      {
        var simple = new AdditionExpression(new Value(2), new Value(3));
        var ep = new ExpressionPrinter();
        ep.Visit(simple);
        Assert.That(ep.ToString(), Is.EqualTo("(2+3)"));
      }

      [Test]
      public void ProductOfAdditionAndValue()
      {
        var expr = new MultiplicationExpression(
          new AdditionExpression(new Value(2), new Value(3)), 
          new Value(4)
          );
        var ep = new ExpressionPrinter();
        ep.Visit(expr);
        Assert.That(ep.ToString(), Is.EqualTo("(2+3)*4"));
      }
    }
  }
}